### PR TITLE
Add correction candidates to FML inspector

### PR DESCRIPTION
**Pull request by jhugman**
Relates to [ EXP-4154](https://mozilla-hub.atlassian.net/browse/EXP-4154).

According to [The Big O of Code Reviews](https://www.egorand.dev/the-big-o-of-code-reviews/), this is a O(_n_) change.

This PR adds error correction candidates API to the errors exposed by the Feature Inspector.

It builds on [several PRs](https://github.com/pulls?q=is%3Apr+author%3Ajhugman+archived%3Afalse+%22relates+to+exp-4154%22+sort%3Acreated-asc):

- #6003
- #6006
- #6007
- #6030
- #6034
- #6037
- #6038

This PR can be sliced in multiple ways:

- In terms of commits:
  1. the first adds the plumbing and the replacements for raw strings.
  2. the second adds replacements for other types (booleans, numbers, objects and lists).
- in terms of lines of code
  - generating the correction candidates
  - adding API surface area.
  - tests.

It's mostly tests.

Context: this is to drive corrections like [this mockup here](https://codemirror.net/try/?c=aW1wb3J0IHtzeW50YXhUcmVlfSBmcm9tICJAY29kZW1pcnJvci9sYW5ndWFnZSI7CmltcG9ydCB7bGludGVyLCBsaW50R3V0dGVyfSBmcm9tICJAY29kZW1pcnJvci9saW50IjsKaW1wb3J0IHtiYXNpY1NldHVwLCBFZGl0b3JWaWV3fSBmcm9tICJjb2RlbWlycm9yIjsKaW1wb3J0IHtqYXZhc2NyaXB0fSBmcm9tICJAY29kZW1pcnJvci9sYW5nLWphdmFzY3JpcHQiOwoKY29uc3QgYWN0aW9ucyA9IFsiQUxXQVlTIiwgIlVTRVJfRU5fU1BFQUtFUiIsICJVU0VSX0RFX1NQRUFLRVIiXS5tYXAoKHMsIGkpID0+IHsKICBjb25zdCBpbnNlcnQgPSBgL2RpZ2l0LSR7aX0vYAogIHJldHVybiB7CiAgICAgICAgICBuYW1lOiBzLAogICAgICAgICAgYXBwbHkodmlldywgZnJvbSwgdG8pIHsgdmlldy5kaXNwYXRjaCh7Y2hhbmdlczoge2Zyb20sIHRvLCBpbnNlcnR9fSk7IH0KICAgICAgICB9Owp9KTsKCiAgY29uc3QgcmVnZXhwTGludGVyID0gbGludGVyKHZpZXcgPT4gewogICAgbGV0IGRpYWdub3N0aWNzID0gW107CiAgICBzeW50YXhUcmVlKHZpZXcuc3RhdGUpLmN1cnNvcigpLml0ZXJhdGUobm9kZSA9PiB7CiAgICAgIGlmIChub2RlLm5hbWUgPT0gIk51bWJlciIpIHsKICAgICAgICBkaWFnbm9zdGljcy5wdXNoKHsKICAgICAgICBmcm9tOiBub2RlLmZyb20sCiAgICAgICAgdG86IG5vZGUudG8sCiAgICAgICAgc2V2ZXJpdHk6ICJlcnJvciIsCiAgICAgICAgbWVzc2FnZTogIkludmFsaWQgdmFsdWUgZm9yIFRyaWdnZXJOYW1lOyBkaWQgeW91IG1lYW4iLAogICAgICAgIGFjdGlvbnMKICAgICAgfSk7fQogICAgfSk7CiAgICByZXR1cm4gZGlhZ25vc3RpY3MKICB9KTsKCiAgLy8hZWRpdG9yCgogIGNvbnN0IG9iaiA9IHsgCiAgICBtZXNzYWdlczogewogICAgICAie2V4cGVyaW1lbnR9LWVuIjogewogICAgICBzdXJmYWNlOiAiaG9tZXNjcmVlbiIsCiAgICAgIAogICAgICB0cmlnZ2VyOiBbMV0sCiAgICAgIAoKICAgICAgdGV4dDogIllvdSBjYW4gY2hhbmdlIHlvdSBzZWFyY2ggZW5naW5lISIsCiAgICAgICJidXR0b24tbGFiZWwiOiAiU3VyZSwgd2h5IG5vdCIsCiAgICAgIGFjdGlvbjogIk9QRU5fU0VUVElOR1NfU0VBUkNIIgogICAgfQogICAgfQogIH07CiAgY29uc3QgZG9jID0gSlNPTi5zdHJpbmdpZnkob2JqLCB1bmRlZmluZWQsIDIpOwoKICBuZXcgRWRpdG9yVmlldyh7CiAgICBkb2MsCiAgICBleHRlbnNpb25zOiBbCiAgICAgIGJhc2ljU2V0dXAsCiAgICAgIGphdmFzY3JpcHQoKSwKICAgICAgbGludEd1dHRlcigpLAogICAgICByZWdleHBMaW50ZXIKICAgIF0sCiAgICBwYXJlbnQ6IGRvY3VtZW50LmJvZHkKICB9KTsKCg==):

image

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/support/nimbus-fml/src/client/inspector.rs [hunk 1]
```diff
@@ -406,3 +408,481 @@ mod unit_tests {
...
+            err.corrections.len(),
+            "No corrections for {input}: {err:?}"
+        );
+
+        for correction in &err.corrections {
+            let input = perform_correction(lines, &err.error_span, correction);
+            let err = inspector.get_first_error(input.clone());
+            assert_eq!(None, err, "Error found in {input}");
+        }
+    }
```
Comment (jhugman) [inline components/support/nimbus-fml/src/client/inspector.rs:461]: This method is the heart of these tests.

These tests test all the layers at the same time, so they are less brittle in the face of changing implementation details.

File: components/support/nimbus-fml/src/client/inspector.rs [hunk 2]
```diff
@@ -406,3 +408,481 @@ mod unit_tests {
...
+
+        try_correcting_single_error(
+            &inspector,
+            &[
+                // 012345678901234567890123
+                r#"{"#,                             // 0
+                r#"  "map": { "invalid": false }"#, // 1
+                r#"}"#,                             // 2
+            ],
+        );
```
Comment (jhugman) [inline components/support/nimbus-fml/src/client/inspector.rs:733]: This was the test that prompted #6030 .

File: components/support/nimbus-fml/src/editing/error_converter.rs [hunk 3]
```diff
@@ -86,15 +98,55 @@ impl ErrorConverter<'_> {
...
+
+        let mut candidates = Vec::with_capacity(strings.len() + placeholders.len());
+        for s in &strings {
+            candidates.push(CorrectionCandidate::string_replacement(s));
+        }
+        for s in &placeholders {
+            candidates.push(CorrectionCandidate::literal_replacement(s));
+        }
+        candidates
+    }
```
Comment (jhugman) [inline components/support/nimbus-fml/src/editing/error_converter.rs:120]: This is the entry point for where corrections are generated from the error.

**Review by jhugman — COMMENTED** [review]
Self-review and comments.

File: components/support/nimbus-fml/src/editing/values_finder.rs [hunk 4]
```diff
@@ -27,13 +27,29 @@ impl<'a> ValuesFinder<'a> {
...
             TypeRef::Enum(type_name) => self.get_enum_values(type_name),
             _ => Default::default(),
         }
     }
+
+    #[allow(dead_code)]
+    #[cfg(feature = "client-lib")]
+    pub(crate) fn all_placeholders(&self, type_ref: &TypeRef) -> BTreeSet<String> {
+        let strings: &[&str] = match type_ref {
+            TypeRef::Boolean => &["true", "false"],
```
[the remark this answers is not in the document]
Comment (jeddai) [reply components/support/nimbus-fml/src/editing/values_finder.rs:42]: Agreed

**Review by jeddai — APPROVED** [review]
This is sick!!

**Pull request merged**